Document Iceberg time travel & snapshot queries (#10515)
trinodb/trino: docs/src/main/sphinx/connector/iceberg.rst

    DELETE FROM iceberg.testdb.customer_orders
    WHERE country = 'US' AND customer = 'Freds Foods'

Snapshots
---------

Iceberg supports a "snapshot" model of data, where table snapshots are

> identified by an snapshot IDs.

Author: Modified to _snapshots are identified by a snapshot ID_

The connector provides a system table containing snapshot information for every
Iceberg table. Snapshots are identified by ``BIGINT`` snapshot IDs.
You can find the snapshot IDs for table ``customer_orders`` by running the
following command::

    SELECT snapshot_id FROM iceberg.testdb."customer_orders$snapshots" ORDER BY committed_at DESC

Snapshot queries
^^^^^^^^^^^^^^^^

The snapshots of an Iceberg table can be queried by using the following command::

   SELECT * FROM iceberg.testdb."customer_orders@8954597067493422955"

In the command above, the special notation ``"customer_orders@8954597067493422955"``
is used to point to a specific snapshot of the ``customer_orders`` Iceberg table.

> The snapshot queries can be also used with special tables::

Reviewer: Nit: maybe `Querying by snapshot ID can also be used with the other [special] metadata tables`.

    DELETE FROM iceberg.testdb.customer_orders
    WHERE country = 'US' AND customer = 'Freds Foods'

Snapshots
---------

Iceberg supports a "snapshot" model of data, where table snapshots are
identified by an snapshot IDs.

> The connector provides a system snapshots table for each Iceberg table.  Snapshots are

Reviewer 2: pre-existing: "system snapshots table" reads a bit weird. May be rephrase to "a system table containing snapshot information for every iceberg table"

> identified by BIGINT snapshot IDs.  You can find the snapshot IDs for table

Reviewer 2: pre-existing: quote BIGINT

``customer_orders`` by running the following command::

>     SELECT snapshot_id FROM iceberg.testdb."customer_orders$snapshots" ORDER BY committed_at

Reviewer 2: may be keep `DESC` as it's a little bit easier to see when someone is playing around

Snapshot queries
^^^^^^^^^^^^^^^^

> The snapshots of an Iceberg tables can be easily queried by using the following command::

Reviewer 3: an Iceberg table can be queried by

   SELECT * FROM iceberg.testdb."customer_orders@8954597067493422955"

> In the command above, the special notation ``"customer_orders@8954597067493422955"`` is used to point to a

Reviewer 3: wrap at 80 columns ideally

specific snapshot of the ``customer_orders`` Iceberg table.

> 

Reviewer 3: remove empty line

The snapshot queries can be also used with special tables::

> 

Reviewer 2: nit: can we avoid double newlines for uniformity in this file?

     SELECT * FROM iceberg.testdb."customer_orders$data@8954597067493422955"
     SELECT * FROM iceberg.testdb."customer_orders$partitions@8954597067493422955"
     SELECT * FROM iceberg.testdb."customer_orders$manifests@8954597067493422955"
     SELECT * FROM iceberg.testdb."customer_orders$files@8954597067493422955"

Rolling back to a previous snapshot
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> You can find the latest snapshot ID for table ``customer_orders`` by running the following command::

Reviewer 3: You can determine the snapshot IDs for a table in the hidden metadata table ``$snapshots``. Use this table to determine the latest snapshot ID with a query similar to the following example::

Tables using v2 of the Iceberg specification support deletion of individual rows
by writing position delete files.

Snapshots
---------

Iceberg supports a "snapshot" model of data, where table snapshots are
identified by a snapshot ID.

The connector provides a system table containing snapshot information for every
Iceberg table. Snapshots are identified by ``BIGINT`` snapshot IDs.
You can find the snapshot IDs for table ``customer_orders`` by running the
following command::

    SELECT snapshot_id FROM iceberg.testdb."customer_orders$snapshots" ORDER BY committed_at DESC

Rolling back to a previous snapshot
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

You can determine the snapshot IDs for a table in the hidden metadata table ``$snapshots``.

> Use this table to determine the latest snapshot ID with a query similar to the following example::

Reviewer 4: Maybe:

Use the \`\`$snapshots\`\` system table to determine the latest snapshot ID like in this example query::

Tables using v2 of the Iceberg specification support deletion of individual rows
by writing position delete files.

Snapshots
---------

Iceberg supports a "snapshot" model of data, where table snapshots are
identified by a snapshot ID.

The connector provides a system table containing snapshot information for every
Iceberg table. Snapshots are identified by ``BIGINT`` snapshot IDs.

> You can find the snapshot IDs for table ``customer_orders`` by running the

Reviewer 4: for the table \`\`customer_orders\`\` -> for the \`\`customer_orders\`\` table ?

> following command::

Reviewer 5: ```suggestion
following query::
```

    SELECT snapshot_id FROM iceberg.testdb."customer_orders$snapshots" ORDER BY committed_at DESC

> Time travel queries

Reviewer 5: I feel like we'd probably want to label this "querying historical data" or something to that effect. "Time travel" sounds a lot more exciting, but it's less precise. I think we've generally avoided using the term time travel in release notes and user-facing docs/communication.

Author: "Time travel" is the term that got established for this kind of functionality.
If we look at Iceberg , Delta , Hudi , Snowflake on their website - everywhere is used this term for this kind of functionality.

^^^^^^^^^^^^^^^^^^^

> The connector offers the ability to query historical data (i.e. the state of the table
> as it was in the previous day/week containing data which in the meantime has been
> modified or deleted) corresponding to a previous snapshot of an Iceberg table.

Reviewer 5: ```suggestion
The connector offers the ability to query historical data. This allows you to query the
table as it was when a snapshot of the table was made, even if the data has since been
modified or deleted.
```

The historical data of the table can be retrieved by specifying the
snapshot identifier corresponding to the version of the table that
needs to be retrieved::

   SELECT * FROM iceberg.testdb.customer_orders FOR VERSION AS OF 8954597067493422955

A different approach of retrieving historical data is to specify

> a point in time in the past (e.g. : a day/week ago)::

Reviewer 5: ```suggestion
a point in time in the past (such as a day or week ago)::
```

Tables using v2 of the Iceberg specification support deletion of individual rows
by writing position delete files.

Snapshots
---------

Iceberg supports a "snapshot" model of data, where table snapshots are
identified by a snapshot ID.

The connector provides a system table exposing snapshot information for every
Iceberg table. Snapshots are identified by ``BIGINT`` snapshot IDs.
For example, you could find the snapshot IDs for the ``customer_orders`` table
by running the following query::

    SELECT snapshot_id FROM iceberg.testdb."customer_orders$snapshots" ORDER BY committed_at DESC

Time travel queries
^^^^^^^^^^^^^^^^^^^

The connector offers the ability to query historical data.
This allows you to query the table as it was when a previous snapshot
of the table was taken, even if the data has since been modified or deleted.

The historical data of the table can be retrieved by specifying the
snapshot identifier corresponding to the version of the table that
needs to be retrieved::

>    SELECT * FROM iceberg.testdb.customer_orders FOR VERSION AS OF 8954597067493422955

Reviewer 3: Break into multiple lines

A different approach of retrieving historical data is to specify

> a point in time in the past (such as a day or week ago). The latest snapshot

Reviewer 3: ```suggestion
a point in time in the past, such as a day or week ago. The latest snapshot
```

> of the table taken before or at specified timestamp in the query will be

Reviewer 3: ```suggestion
of the table taken before or at specified the timestamp in the query is
```

internally used for providing the previous state of the table::

>    SELECT * FROM iceberg.testdb.customer_orders FOR TIMESTAMP AS OF TIMESTAMP '2022-03-23 09:59:29.803 Europe/Vienna'

Reviewer 3: Break into multiple lines

Rolling back to a previous snapshot
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Use the ``$snapshots`` metadata table to determine the latest snapshot ID of the table like in the following query::

>     SELECT snapshot_id FROM iceberg.testdb."customer_orders$snapshots" ORDER BY committed_at DESC LIMIT 1

Reviewer 3: Break into multiple lines (probably in separate PR or commit)
